dix/req.c

```diff
@@ -2153,7 +2224,26 @@ static int dix_cas_rops_alloc(struct m0_dix_req *req)
 	if (cas_rop_tlist_is_empty(&rop->dg_cas_reqs))
 		return M0_ERR(-EIO);
 
+	if (dtx != NULL) {
+		M0_ASSERT(!req->dr_is_meta);
+		M0_ASSERT(M0_IN(req->dr_type, (DIX_PUT, DIX_DEL)));
+		rc = m0_dtx0_open(dtx, cas_rop_tlist_length(&rop->dg_cas_reqs));
+		if (rc != 0)
+			goto end;
+
+		i = 0; /* reset it to enumerate dtx parcticipants */
+	}
+
 	m0_tl_for(cas_rop, &rop->dg_cas_reqs, cas_rop) {
+		if (dtx != NULL) {
+			cas_rop->crp_pa_idx = i++;
+			cas_svc = pc->pc_dev2svc[cas_rop->crp_sdev_idx].pds_ctx;
+			M0_ASSERT(cas_svc->sc_type == M0_CST_CAS);
+			rc = m0_dtx0_fid_assign(dtx, cas_rop->crp_pa_idx,
+						&cas_svc->sc_fid);
+			if (rc != 0)
+				goto end;
```

Comment:
Please note, the dtx is open. 
Do we need to do cleanup as part of error handling?

Reply:
In this particular place we do not need to cleanup the dtx (because it is owned by the client operation object but not by the dix request), and the client part (idx/idx_dix) must handle that at the moment when idx_op_fail is called. Although idx_op_fail is something that we were going to get to in the consequent patches, let's mark this missing part with TODO.